src/python/pants/engine/internals/native.py

```diff
@@ -34,35 +50,85 @@
 
     TODO: These could be implemented in Rust in `externs.rs` via the cpython API.
     """
 
-    def __init__(self, lib):
+    def __init__(self, lib: ModuleType):
         self.lib = lib
 
+    # NB: This flag is exposed for testing error handling in CFFI methods. This should never be set
+    # in normal pants usage.
     _do_raise_keyboardinterrupt = bool(os.environ.get("_RAISE_KEYBOARDINTERRUPT_IN_EXTERNS", False))
 
-    def is_union(self, input_type):
+    def is_union(self, input_type: Type) -> bool:
         """Return whether or not a type is a member of a union."""
         # NB: This check is exposed for testing error handling in CFFI methods. This code path should
         # never be active in normal pants usage.
         return union.is_instance(input_type)
 
-    def create_exception(self, msg):
-        """Given a utf8 message string, create an Exception object."""
-        return Exception(msg)
+    def create_value_error(self, msg: str) -> IntrinsicError:
+        """Given a utf8 message string, create an exception signalling a value error."""
+        return IntrinsicError(msg)
+
+    def create_type_error(self, msg: str) -> IncorrectProductError:
+        """Given a utf8 message string, create an exception signalling a type error."""
+        return IncorrectProductError(msg)
+
+    # A specification for how the native engine interacts with @rule coroutines:
+    # - coroutines may await on a `Get` or a tuple of `Get`.
+    # - we will send back a single `Any` or a tuple of `Any`, depending upon the variant of `Get`.
+    # - a coroutine will eventually return a single `Any`.
+    RuleSend = Union[Any, Tuple[Any, ...]]
+    RuleYield = Union[Get, Tuple[Get, ...]]
+    RuleCoroutine = Coroutine[RuleSend, RuleYield, Any]
+
+    @overload
+    def _send_to_coroutine(self, func: RuleCoroutine, arg: RuleSend) -> RuleYield:
+        ...
+
+    @overload
+    def _send_to_coroutine(self, func: RuleCoroutine, arg: Exception) -> RuleYield:
+        ...
+
+    def _send_to_coroutine(self, func: RuleCoroutine, arg: Union[RuleSend, Exception]) -> RuleYield:
+        """We can send a coroutine a single object, a tuple, or an exception.
+
+        We send coroutines inputs as the result of an `await Get(...)`. An exception occurs here
+        when any @rule or intrinsic called within an `await Get(...)` raises an exception instead of
+        returning a value.
+        """
+        if isinstance(arg, Exception):
+            # Coroutine.throw(self, typ, val=None, tb=None) will raise `typ` (the first positional
+            # argument) as the exception if `val` and `tb` are None (as they are here).
+            return func.throw(arg)  # type: ignore[arg-type]
+        return func.send(arg)
 
     def generator_send(
-        self, func, arg
-    ) -> Union[PyGeneratorResponseGet, PyGeneratorResponseGetMulti, PyGeneratorResponseBreak]:
+        self,
+        func: RuleCoroutine,
+        arg: Union[RuleSend, Exception],
+    ) -> Union[
```

Comment:
Seems like it would make sense to define a name for this long `Union` and use that name in the actual function signature

Reply:
Will do!